src/app/public/modules/inline-delete/inline-delete-adapter.service.ts

```diff
@@ -0,0 +1,134 @@
+import {
+  Injectable,
+  Renderer2,
+  RendererFactory2
+} from '@angular/core';
+
+/**
+ * @internal
+ */
+@Injectable()
+export class SkyInlineDeleteAdapterService {
+
+  private element: HTMLElement;
+  private focussableElements: HTMLElement[];
+  private parentEl: HTMLElement;
+  private parentElUnlistenFn: Function;
+  private renderer: Renderer2;
+
+  constructor(
+    private rendererFactory: RendererFactory2
+  ) {
+    this.renderer = this.rendererFactory.createRenderer(undefined, undefined);
+  }
+
+  public clearListeners(): void {
+    this.parentElUnlistenFn();
+  }
+
+  public setEl(element: HTMLElement): void {
+    this.element = element;
+    this.parentEl = element.parentElement;
+    this.parentElUnlistenFn = this.renderer.listen(this.parentEl, 'focusin',
+      (event: KeyboardEvent) => {
+        const target: any = event.target;
+        if (!this.element.contains(target)) {
+          event.preventDefault();
+          event.stopPropagation();
+          event.stopImmediatePropagation();
+
+          target.blur();
+          this.focusNextElement(target, this.isShift(event), this.parentEl);
+        }
+      });
+  }
+
+  private focusNextElement(targetElement: HTMLElement, shiftKey: boolean, busyEl: Element): void {
+    const focussable = this.getFocussableElements();
+
+    // If shift tab, go in the other direction
+    const modifier = shiftKey ? -1 : 1;
+
+    // Find the next navigable element that isn't waiting
+    const startingIndex = focussable.indexOf(targetElement);
+    let curIndex = startingIndex + modifier;
+    while (focussable[curIndex] && this.isElementHiddenOrCovered(focussable[curIndex])) {
+      curIndex += modifier;
+    }
+
+    if (focussable[curIndex] && !this.isElementHiddenOrCovered(focussable[curIndex])) {
+      focussable[curIndex].focus();
+    } else {
+      // Try wrapping the navigation
+      curIndex = modifier > 0 ? 0 : focussable.length - 1;
+      while (
+        curIndex !== startingIndex &&
+        focussable[curIndex] &&
+        this.isElementHiddenOrCovered(focussable[curIndex])
+      ) {
+        curIndex += modifier;
+      }
+
+      /* istanbul ignore else */
+      /* sanity check */
+      if (focussable[curIndex] && !this.isElementHiddenOrCovered(focussable[curIndex])) {
+        focussable[curIndex].focus();
+      } else {
+        // No valid target, wipe focus
+        // This should never happen in practice due to the multiple inline delete buttons
+        if (document.activeElement && (document.activeElement as any).blur) {
+          (document.activeElement as any).blur();
+        }
+        document.body.focus();
+      }
+    }
+
+    // clear focussableElements list so that if things change between tabbing we know about it
+    this.focussableElements = undefined;
+  }
+
+  private getFocussableElements(): HTMLElement[] {
```

Comment:
Can you use reuse skyux-core's adapter service for a lot of this logic?
https://github.com/blackbaud/skyux-core/blob/92fbcdb07ed4f8448cde87a97c7922f76a95605e/src/app/public/modules/adapter-service/adapter.service.ts

Reply:
Done